### PR TITLE
[CI Implementation] - Run unit tests

### LINKED ISSUE #54: [CI] - CI Build
As a car owner, I want the peace of mind that all new code is automatically built by a CI framework.

### PR DESCRIPTION
**Pull request by Gustmill**
- This resolves #54 and #57 
- Travis configuration
- Run unit tests in Travis
- Implemented build status in the readme file

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+addons:
+  apt:
+    sources:
+#      - ppa:chris-lea/redis-server
+    packages:
+#      - redis-tools
+#      - redis-server
+services:
+  - postgresql
+
```
Comment (leithhobson) [inline .travis.yml:35]: Looks good!
Now it is just to get the C++ Unit tests implemented

**Review by leithhobson — APPROVED** [review]

File: Car/TeamNull SmartCar/src/manager/MovementManager.cpp [hunk 2]
```diff
@@ -1,63 +1,227 @@
...
+boolean colisionBeingAvoided = false;
+boolean obstacleBeingAvoided = false;
+boolean obstacleAvoidanceTurnedRight = false;
+boolean obstacleAvoidanceTurnedLeft = false;
+long timeObstacleWasAvoided = millis();
+long timereturnToRouteBegan = millis();
 
-void setDesiredVehicleSpeed (int speed) {
+//This is a helper method for logging purposes only
+String booleanToString(boolean boolean)
```
Comment (platisd) [inline Car/TeamNull SmartCar/src/manager/MovementManager.cpp:18]: It's probably not the best idea to call a variable after a type. Would you ever write `int int`? :)

File: Car/TeamNull SmartCar/src/util/HardwareUtil.cpp [hunk 3]
```diff
@@ -23,6 +69,22 @@ void setAngle(int angle) {
...
+  // Serial.print("Left: ");
+  // Serial.println(left);
+  // Serial.print("Right: ");
+  // Serial.println(right);
+  // Serial.print("Combined: ");
+  // Serial.println((left+right)/2);
+  
+
+  return (((left+right)/2)*100);
+}
```
Comment (platisd) [inline Car/TeamNull SmartCar/src/util/HardwareUtil.cpp:86]: You can just use `car.getSpeed`

File: Car/TeamNull SmartCar/src/util/HardwareUtil.cpp [hunk 4]
```diff
@@ -4,17 +4,63 @@
...
+//     // correct pulses:centimeters ratio for your odometer
+//     // that you should use when initializing the odometer
+//     // in its constructor
+//     Serial.print("Left: ");
+//     Serial.println(leftOdometer.getDistance());
+//     Serial.print("Right: ");
+//     Serial.println(rightOdometer.getDistance());
+
+//     delay(100);
+// }
```
Comment (platisd) [inline Car/TeamNull SmartCar/src/util/HardwareUtil.cpp:55]: If you aren't using some code just remove it

File: Car/TeamNull SmartCar/src/util/HardwareUtil.cpp [hunk 5]
```diff
@@ -58,58 +124,62 @@ void initialiseSensors() {
...
+    pinMode(rearSensorPin, OUTPUT);
+    digitalWrite(rearSensorPin, LOW);
+
+    delay(1500);
+    pinMode(rearSensorPin, INPUT);
+  }
   // rearSensor.setAddress((uint8_t)02);
 
+  SR04 frontRight(FR_TRIGGER_PIN, FR_ECHO_PIN, MAX_DISTANCE);
+  SR04 frontLeft(FL_TRIGGER_PIN, FL_ECHO_PIN, MAX_DISTANCE);
```
Comment (platisd) [inline Car/TeamNull SmartCar/src/util/HardwareUtil.cpp:173]: Why are you redeclaring these?

File: Car/TeamNull SmartCar/src/util/HardwareUtil.cpp [hunk 6]
```diff
@@ -23,6 +69,22 @@ void setAngle(int angle) {
...
+  // Serial.println("Speed:");
+  // Serial.print("Left: ");
+  // Serial.println(left);
+  // Serial.print("Right: ");
+  // Serial.println(right);
+  // Serial.print("Combined: ");
+  // Serial.println((left+right)/2);
+  
+
+  return (((left+right)/2)*100);
```
Comment (platisd) [inline Car/TeamNull SmartCar/src/util/HardwareUtil.cpp:85]: If you use a `DistanceCar` then you can do `car.getSpeed` instead.

File: Car/TeamNull SmartCar/src/util/HardwareUtil.cpp [hunk 7]
```diff
@@ -121,13 +191,16 @@ int getFrontDistance() {
 }
 
 int getLeftFrontDistance() {
-  return 100;
+  return frontLeft.getMedianDistance(2);
```
Comment (platisd) [inline Car/TeamNull SmartCar/src/util/HardwareUtil.cpp:194]: What's the point of getting the median distance of two measurements? You'd need at least three for a median to make sense.

**Review by platisd — CHANGES_REQUESTED** [review]

**Comment by platisd** [comment]
Approved already? Did you **really** code review this? :scream:

**Comment by leithhobson** [comment]
> Approved already? Did you **really** code review this? scream

I really code reviewed the CI part of this PR
My and Martin's commits related to the car, on the other hand, are NOT final, but are already in this branch, and so they were ignored for the purpose of reviewing the code, as it would be pointless to review code that is known to be incomplete

That said, your comments had not come up in my feed before I merged, although they now show as being before the merge

**Comment by platisd** [comment]
If this pull request is about setting up the CI, why are you also including unrelated and (to make things worse) incomplete code?

> That said, your comments had not come up in my feed before I merged, although they now show as being before the merge

Probably because I started reviewing while this was still not merged.

File: Car/TeamNull SmartCar/src/manager/MovementManager.cpp (hunk 2 above)
Comment (leithhobson) [reply Car/TeamNull SmartCar/src/manager/MovementManager.cpp:18]: I have seen String string many times, however, I've now renamed this variable in my own branch

File: Car/TeamNull SmartCar/src/util/HardwareUtil.cpp (hunk 4 above)
Comment (leithhobson) [reply Car/TeamNull SmartCar/src/util/HardwareUtil.cpp:55]: Removed (in my branch)

File: Car/TeamNull SmartCar/src/util/HardwareUtil.cpp (hunk 5 above)
Comment (leithhobson) [reply Car/TeamNull SmartCar/src/util/HardwareUtil.cpp:173]: That would be PEBCAK :/ resolved (in my branch)

File: Car/TeamNull SmartCar/src/util/HardwareUtil.cpp (hunk 7 above)
Comment (leithhobson) [reply Car/TeamNull SmartCar/src/util/HardwareUtil.cpp:194]: We had some issues with code getting stuck 'somewhere' and so this was lowered from 5 to 2 for testing, but yes, it is intended to be higher

**Comment by leithhobson** [comment]
> If this pull request is about setting up the CI, why are you also including unrelated and (to make things worse) incomplete code?
> 
> > That said, your comments had not come up in my feed before I merged, although they now show as being before the merge
> 
> Probably because I started reviewing while this was still not merged.

I understand, I was just saying that I didn't ignore your comments intentionally

**Comment by leithhobson** [comment]
> If this pull request is about setting up the CI, why are you also including unrelated and (to make things worse) incomplete code?

Because when one makes a PR from a branch, it includes all commits on the branch, and I don't think anyone on the team knows how to remove a commit that has been merged in from another branch. There was a merge in order to get an update to the pom.xml from my personal branch

**Comment by platisd** [comment]
> > If this pull request is about setting up the CI, why are you also including unrelated and (to make things worse) incomplete code?
> 
> Because when one makes a PR from a branch, it includes all commits on the branch, and I don't think anyone on the team knows how to remove a commit that has been merged in from another branch. There was a merge in order to get an update to the pom.xml from my personal branch

:exploding_head:

I see. Time to let the team know about [cherry pick](https://git-scm.com/docs/git-cherry-pick) then! :)

**Comment by leithhobson** [comment]
> I see. Time to let the team know about [cherry pick](https://git-scm.com/docs/git-cherry-pick) then! :)

Thank you, we will try to improve our practises in the upcoming sprints

**Pull request merged**